specs/hierarchical_consensus.md

```diff
@@ -0,0 +1,1086 @@
+
+Validators `CrossMsgPool`s also listen for new `BottomUpMsgsMeta` being included in their subnet SCA. When a new `CrossMsgsMeta` appear in `BottomUpMsgsMeta` (after the committment of a checkpoint), the `CrossMsgPool` checks if `CrossMsgsMeta.Nonce > AppliedBottomUpNonce` to see if it includes cross-msgs that haven't been executed yet. If this is the case, the `CrossMsgPool`:
+- Gets all `CrossMsgsMeta` in `BottomUpMsgsMeta` with `Nonce > AppliedBottomUpNonce`.
+- Gets `CrossMsgMeta.Cid` and makes a request to the subnet content resolution protocol to resolve the `CrossMsgs` behind that `CrossMsgMeta`. These requests are directed to the subnet in `Source` and they resolve the CID from the subnet's `CrossMsgsRegistry`. 
+    - If the resolved `CrossMsgs` only includes elements in the `Msgs` field, they can be directly proposed in the next block for their execution.
+    - If this is not the case, and `CrossMsgs` includes in its `Meta` field `CrossMsgMetas` from its children, then these `CrossMsgsMeta` need to be resolved recursively until all the `CrossMsgsMeta` have been resolved to their underlaying messages successfully.
+- Then, as it happened for top-down messages, when the `CrossMsgPool`  has all the bottom-up messages to be applied, it waits for a `FinailityThreshold` before proposing the message to be sure that the checkpoint commitment can be considered final, and all the resolved cross-messages are proposed for their inclusion and subsequent execution. 
+
+![](https://hackmd.io/_uploads/Bkrzj2uF5.png)
+
+### Path messages
+Path messages are propagated and executed as a combination of bottom-up and top-down messages according the path they need to traverse in the hierarchy. Let's consider a message from `/root/f0100/f0200` to `/root/f0100/f0300`. This message is propagated as a set of bottom-up message up to the closes common parent (`/root/f0100` in our example). When the checkpoint including the cross-message from `/root/f0100/f0200` arrives to `/root/f0100`, the `CrossMsgMeta` is resolved, and from there on, the message is propagated as a top-down message from the closest common parent to the destination (in this case, from `/root/f0100` to `/root/f0100/f0300`),
+
+#### Errors propagating cross-net messages
+If at any point the propagation or execution of a cross-msg fails (either because the message runs out of gas in-transit; because the message is invalid in the destination subnet; or because the `CrossMsgMeta.Cid` can't be resolved successfully), the message is discarded, and a error is tracked for the cross-message.
```

Comment:
```suggestion
If at any point the propagation or execution of a cross-msg fails (either because the message runs out of gas in-transit; because the message is invalid in the destination subnet; or because the `CrossMsgMeta.Cid` can't be resolved successfully), the message is discarded, and an error is tracked for the cross-message.
```

What does it mean to "track" an error?